src/lib.rs

```diff
@@ -1,576 +1,333 @@
-use std::cmp::Ordering;
-use std::collections::BTreeMap;
-use std::time::SystemTime;
+use std::{
+    ops::Add,
+    time::{Duration, SystemTime},
+};
 
 // TODO: better error management
-use anyhow::{ensure, Context, Result};
-use ed25519_dalek::ed25519::signature::Signer;
-use ed25519_dalek::{Signature, SigningKey, VerifyingKey};
-use rand::{CryptoRng, Rng};
-use serde::de::DeserializeOwned;
-use serde::{Deserialize, Serialize};
-
-// TODO: what's the smalles secure value?
-pub const NONCE_SIZE: usize = 12;
+use anyhow::{bail, ensure, Context, Result};
+use ed25519_dalek::{
+    ed25519::signature::Signer, Signature, SigningKey, VerifyingKey, SIGNATURE_LENGTH,
+};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
 
 pub const VERSION: u8 = 1;
 
-pub trait Capability:
-    std::fmt::Debug + Sized + Serialize + DeserializeOwned + PartialEq + Eq + Clone
-{
+/// We allow arbitrary capabilities.
+///
+/// An example for a type implementing this trait might be the enum
+/// for some RPC requests.
+///
+/// Capabilities are restricted by "attenuations", which are representations
+/// of predicates on them. `allowed_by_attenuation` is what translates the
+/// predicate representation into an actual predicate.
+///
+/// The `attenutation` must be serializable so it can be used in an rcan,
+/// which is signed.
+pub trait Attenuation: Serialize {
```

Comment:
meh - now that this is only one concept, let's perhaps call it "Capability", and `Capability::can_delegate`.

Reply:
done